Assignment3

Laucalle opened the pull request.

Laucalle committed:
Add files for assignment 3

large_dictionary.txt is used for testing

Laucalle committed:
Create assignment3.cpp

Laucalle committed:
Create large_dictionary.txt

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+		std::ifstream file (filemame);
+  		if (file.is_open()) {
+ 			std::string line;
+    		while (std::getline (file,line)) {
```
Indentation is off here.

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+      			dictionary.push_back(line);
+    		}
+    		file.close();
+  		}
```
else ?

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+	bool isPrefix(const std::string &word) const {
+		bool is_prefix = false;
+		for(int i = 0; i<dictionary.size() && !is_prefix; i++){
+			is_prefix = true;
```
I think lines 25-31 could be greatly simplified using some string function like compare or substr

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+#include <set>
+
+class Dictionary{
+	std::vector<std::string> dictionary;
```
The Dictionary is not the main part of the problem, so fell free to ignore this, but using a unsorted array for the dictionary may be "a bit" slow. Have you tried with a medium dictionary (50k-100k words), and a 4x4 grid ?

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+	void findWord(char **grid, int rows, int columns, const Dictionary &dict, 
+				  std::vector<slot> &path, std::set<std::string> &solution){
+
+		std::string word = getString(grid,rows,columns,path);
```
Having the full path and recomputing the full word each time seems like more work than needed.

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+		bool already_used;
+		for(int i = init_row; i<=end_row; i++){
+			for(int j= init_col; j<=end_col; j++){
+				if(!( i == current_pos.first && j== current_pos.second)){
```
This is not needed, right? The current cell is already in the path.

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+		for(int i = init_row; i<=end_row; i++){
+			for(int j= init_col; j<=end_col; j++){
+				if(!( i == current_pos.first && j== current_pos.second)){
+					already_used = false;
```
Maybe you can use vector:: find instead?

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+
+int main(){
+	Dictionary dict("large_dictionary.txt");
+	char **grid;
```
I think `vector<vector<char>>`  or `vector<string>` are better choices here. Just being able to pass them around and knowing the length is priceless :)
They are also easier to initialise, for example:
`vector<string> grid{"aar", "tcd"};
`

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+		int init_col = current_pos.second-1;
+		int end_col = current_pos.second+1;
+
+		init_col = init_col<0?0:init_col;
```
A question of taste, but some prefer to use `max` for this (and `min` in the other case)

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+	for (std::set<std::string>::iterator it=words.begin(); it!=words.end(); ++it){
+    	std::cout << *it << "\n";
+	}
+}
```
Remember to delete the things you allocated (grid & grid2). Or use vectors :)

ssancho2 commented on assignment3.cpp:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	words = wordfinder.FindWord(grid2, n_rows, n_col, dict);
+	for (std::set<std::string>::iterator it=words.begin(); it!=words.end(); ++it){
```
Easier to use:
for (const string& word : words) { ... }

ssancho2 reviewed (commented):
Good to have real tests with a real dictionary!
Main comments about performance. This of course depends on context, on a real work environment you will measure with real data and see if performance is good enough. In a interview context you need to discuss alternatives with your interviewer, that will probably request something more performant :)

Laucalle committed:
Updated Assignment 3 after comments

In upcoming commits I'll also address:
  - change dictionary  to use a set
  - max/min change from line 75

ssancho2 commented on assignment3/assignment3.cpp:
```diff
@@ -6,191 +7,158 @@
+
+
+
+    for(int i=0; i<grid.size(); i++){
```
Being a `vector<string>` you only need one for loop, and using the iterator for syntax you can do:
```
for (const string &line : grid) {
   std::cout << line << std::endl;
}
```

ssancho2 commented on assignment3/assignment3.cpp:
```diff
@@ -6,191 +7,158 @@
+            }
+            file.close();
+        } else {
+            dictionary.clear();
```
I think it will be good to either inform the user that the file couldn't be open (including the filename in the message), or throw an exception (but this is not used much).
This clear will have no effect, right? We are in the constructor and haven't added anything to the dictionary yet.

ssancho2 commented on assignment3/assignment3.cpp:
```diff
@@ -6,191 +7,158 @@
+        bool is_prefix = false;
+
+        for(int i = 0; i<dictionary.size() && !is_prefix; i++){
+            is_prefix = word.compare(dictionary[i].substr(0,word.size())) == 0;
```
The idea was to use one not necessarily both :)
I think this version of compare is equivalent to using `==`, i.e: 
`word == dictionary[i].substr(0,word.size())`
Or using another form of compare:
`word.compare(0, word.size(), dictionary[i]) == 0`

ssancho2 commented on assignment3/assignment3.cpp:
```diff
@@ -6,191 +7,158 @@
+                  std::vector<slot> &path, std::string &word,
+                  std::set<std::string> &solution){
+
+        // std::string word = getString(grid,rows,columns,path);
```
Please remove commented out code.

ssancho2 commented on assignment3/assignment3.cpp:
```diff
@@ -6,191 +7,158 @@
-	}
+    typedef std::pair<int, int> slot;
+    
+    std::string getString(char **grid, int rows, int columns, 
```
No need for this function anymore, right?

ssancho2 commented on assignment3/assignment3.cpp:
```diff
@@ -6,191 +7,158 @@
+        return word;
+    }
+
+    void findWord(std::vector<std::string> &grid, const Dictionary &dict, 
```
grid can also be `const`

ssancho2 commented on assignment3/assignment3.cpp:
```diff
@@ -0,0 +1,164 @@
+private:
+    typedef std::pair<int, int> slot;
+    
+    std::string getString(char **grid, int rows, int columns, 
```
This is not needed anymore, right?

ssancho2 reviewed (approved):
A few more minor comments.
